libs/coroutines/include/hpx/coroutines/stackless_coroutine.hpp

```diff
@@ -0,0 +1,239 @@
+        using arg_type = thread_state_ex_enum;
+
+        using functor_type =
+            util::unique_function_nonser<result_type(arg_type)>;
+
+        stackless_coroutine(functor_type&& f, thread_id_type id,
+            std::ptrdiff_t stack_size = default_stack_size)
+          : f_(std::move(f))
+          , state_(ctx_ready)
+          , id_(id)
+#if defined(HPX_HAVE_THREAD_PHASE_INFORMATION)
+          , phase_(0)
+#endif
+#if defined(HPX_HAVE_THREAD_LOCAL_STORAGE)
+          , thread_data_(nullptr)
+#else
+          , thread_data_(0)
+#endif
+        {
+        }
+
+        ~stackless_coroutine()
+        {
+#if defined(HPX_HAVE_THREAD_LOCAL_STORAGE)
+            detail::delete_tss_storage(thread_data_);
+#else
+            thread_data_ = 0;
+#endif
+        }
+
+        stackless_coroutine(stackless_coroutine const& src) = delete;
+        stackless_coroutine& operator=(stackless_coroutine const& src) = delete;
+        stackless_coroutine(stackless_coroutine&& src) = delete;
+        stackless_coroutine& operator=(stackless_coroutine&& src) = delete;
+
+        thread_id_type get_thread_id() const
+        {
+            return id_;
+        }
+
+#if defined(HPX_HAVE_THREAD_PHASE_INFORMATION)
+        std::size_t get_thread_phase() const
+        {
+            return phase_;
+        }
+#endif
+        std::size_t get_thread_data() const
+        {
+#if defined(HPX_HAVE_THREAD_LOCAL_STORAGE)
+            if (!thread_data_)
+                return 0;
+            return detail::get_tss_thread_data(thread_data_);
+#else
+            return thread_data_;
+#endif
+        }
+
+        std::size_t set_thread_data(std::size_t data)
+        {
+#if defined(HPX_HAVE_THREAD_LOCAL_STORAGE)
+            return detail::set_tss_thread_data(thread_data_, data);
+#else
+            std::size_t olddata = thread_data_;
+            thread_data_ = data;
+            return olddata;
+#endif
+        }
+
+#if defined(HPX_HAVE_THREAD_LOCAL_STORAGE)
+        detail::tss_storage* get_thread_tss_data(bool create_if_needed) const
+        {
+            if (!thread_data_ && create_if_needed)
+                thread_data_ = detail::create_tss_storage();
+            return thread_data_;
+        }
+#endif
+
+        void rebind(functor_type&& f, thread_id_type id)
+        {
+            HPX_ASSERT(exited());
+
+            f_ = std::move(f);
+            id_ = id;
+
+#if defined(HPX_HAVE_THREAD_PHASE_INFORMATION)
+            phase_ = 0;
+#endif
+#if defined(HPX_HAVE_THREAD_LOCAL_STORAGE)
+            HPX_ASSERT(thread_data_ == nullptr);
+#else
+            HPX_ASSERT(thread_data_ == 0);
+#endif
+            state_ = stackless_coroutine::ctx_ready;
+        }
+
+        void reset()
+        {
+            HPX_ASSERT(exited());
+
+            util::detail::reset_function(f_);
+
+#if defined(HPX_HAVE_THREAD_PHASE_INFORMATION)
+            phase_ = 0;
+#endif
+#if defined(HPX_HAVE_THREAD_LOCAL_STORAGE)
+            detail::delete_tss_storage(thread_data_);
```

Comment:
Shouldn't we apply the tss fixes to the stackless coroutines as well?

Reply:
Well, we could, certainly - even if just for consistency. There is no way the TSS problem can happen for stackless threads, I believe.